Added Github Action to check number of files submitted during PR.

Anubhav-2003 opened the pull request:
<!--
Thanks for submitting a pull request! Please provide enough information so that others can review your pull request.
-->

**What kind of change does this PR introduce?**

<!-- E.g. a bugfix, feature, refactoring, etc… -->
This PR introduces a feature. The feature ensures that every time a new PR is submitted. The number of files submitted
is less than 20. If it exceeds 20, the Github Action Fails.

**Issue Number:**
Fixes #1556

**Did you add tests for your changes?**

<!--Yes or No. Note: Add unit tests or automation tests for your code.-->
No
**Snapshots/Videos:**

<!--Add snapshots or videos wherever possible.-->

**If relevant, did you update the documentation?**
No
<!--Add link to Talawa-Docs.-->

**Summary**

<!-- Explain the **motivation** for making this change. What existing problem does the pull request solve? -->
<!-- Try to link to an open issue for more information. -->

**Does this PR introduce a breaking change?**

<!-- If this PR introduces a breaking change, please describe the impact and a migration path for existing applications. -->

**Other information**

<!--Add extra information about this PR here-->

**Have you read the [contributing guide](https://github.com/PalisadoesFoundation/talawa-api/blob/master/CONTRIBUTING.md)?**
Yes
<!--Yes or No-->

noman2002 commented on .github/workflows/check_files_submitted.py:
```diff
@@ -0,0 +1,88 @@
+    """
+    try:
+        with open(file_name, "r", encoding="utf-8") as file:
+            num_lines = len(file.readlines())
+            if num_lines > 20:
+                print("Number of changed files exceeds 20.")
+                return False
```
why this is set to 20.

Anubhav-2003 replied:
Thank you for your response. It is set to 20,  because in the issue it was written that if the no of files submitted as changed in a single PR exceeds 20. Then it must fail. It was written in the propsed solution.

palisadoes commented:
Please work with @noman2002 on this. He'll be taking over the review of this PR

Anubhav-2003 commented:
@palisadoes ok sir.

palisadoes commented on .github/workflows/pull-request.yml:
```diff
@@ -58,6 +58,37 @@ jobs:
+        with:
+          python-version: '3.11.6'
+
+      - name: Install dependencies
```
This is not required

palisadoes commented on .github/workflows/pull-request.yml:
```diff
@@ -58,6 +58,37 @@ jobs:
+      - name: Get changed files in base branch, compare difference, and execute python script
+        id: changed-files-base
+        run: |
+          git fetch origin $BASE_BRANCH
```
Include this logic within the script. No exporting of output to a file.

palisadoes replied:
Use subprocess.Popen.communicate it gives you a lot of options for getting the output of OS commands and specific error handling
- https://docs.python.org/3/library/subprocess.html#subprocess.Popen.communicate

palisadoes commented on .github/workflows/check_files_submitted.py:
```diff
@@ -0,0 +1,88 @@
+    return True
+
+
+def run_checks():
```
This isn't required. The formatting requirement was for your benefit. It wasn't required for the script to police itself.

palisadoes commented on .github/workflows/check_files_submitted.py:
```diff
@@ -0,0 +1,88 @@
+        print("Usage: python script_name.py <file_to_check>")
+        sys.exit(1)
+
+    file_to_check = sys.argv[1]
```
Use Argparse. It offers a great deal of flexibility

palisadoes reviewed (changes requested):
See comments

Anubhav-2003 commented:
@palisadoes ok sir, I am onto the changes.

Anubhav-2003 commented:
@palisadoes Respected sir, I have tried to implement the changes according to your suggestions. Using subprocess.popen, as well as using argparse to handle Command line arguments. Thank you

palisadoes commented on .github/workflows/check_files_submitted.py:
```diff
@@ -0,0 +1,108 @@
+    pr_branch = args.pr_branch
+
+    changed_files = get_changed_files(base_branch, pr_branch)
+    if len(changed_files) > 20:
```
Get this value from argparse with a default of `20`

palisadoes commented on .github/workflows/check_files_submitted.py:
```diff
@@ -0,0 +1,108 @@
+        None. (Exits with sys.exit() if conditions are met)
+    """
+    parser = argparse.ArgumentParser(description="Check the number of changed files.")
+    parser.add_argument("base_branch", help="Base branch name")
```
Should these be mandatory arguments?

Anubhav-2003 replied:
I think that these arguments should be mandatory, and since these are positional arguments they are inherently mandatory.

palisadoes commented on .github/workflows/pull-request.yml:
```diff
@@ -58,6 +58,31 @@ jobs:
+        run: |
+          git fetch origin $BASE_BRANCH
+          git checkout $BASE_BRANCH
+          python .github/workflows/check_files_submitted.py $BASE_BRANCH pr-branch
```
Add the `base_branch` and `pr_branch` strings here too to make it very clear what is happening

Anubhav-2003 replied:
If I do it using '--', then they will no longer be positional arguments and therefore not mandatory. Therefore the script would become more complicated trying to handle the error cases manually, exceeding more than 130 lines. So I decided to provide more context using comments as to what exactly is happening. 

Let me know if you want it to be handled differently. Thank you.

palisadoes replied:
See the previous comment

palisadoes commented on .github/workflows/check_files_submitted.py:
```diff
@@ -0,0 +1,108 @@
+
+            if git_diff_process.returncode != 0:
+                print(f"Error: {git_diff_error}")
+                return []
```
You have an error condition and this logic will make the script terminate with a normal exit code

palisadoes commented on .github/workflows/check_files_submitted.py:
```diff
@@ -0,0 +1,108 @@
+            changed_files = git_diff_output.splitlines()
+            return changed_files
+    except Exception as e:
+        print(f"Error: {e}")
```
You have an error condition and this logic will make the script terminate with a normal exit code

palisadoes reviewed (changes requested):
See comments

Anubhav-2003 commented:
@palisadoes Respected sir, I have tried to make the changes requested. 
1.)  Added the value 20 as default for max_files in argparser. And used that in the if condition.
2.) Made the process terminate with an appropriate exit code, if errors are encountered.
3.) Added detailed context to the job step performing the task of finding diff between the branches.

Let me know if you need further modification. Thank you.

xoldyckk commented:
is this related to git merge vs rebase?

palisadoes commented:
> is this related to git merge vs rebase?

Based on the YAML its comparing the `develop` base branch with the proposed changes in the branch of the author of the PR

Anubhav-2003 commented:
@palisadoes Respected sir, 
I have made the followoing changes:
1. Removed positional argument status, and added flag (required=True) in argparser.
2. Removed comments and explicitly mentioned the named parameters using "--" for better clarity.

I have tried to implement it the way instructed by you. If you want any further changes, please let me know sir.
Thank You.

xoldyckk commented:
@palisadoes i'm guessing people are doing `git merge main` in their feature branches and pulling changes from an inactive older branch?

not sure if limiting the amount of files in a PR is a good way to do this

since `main` is the default branch in github, most contributers/visitors when they view the repo through their devices they are looking at the main branch, i would suggest changing role of the main branch to the one that is actively worked on and moving the current main branch to something named `stable/prod` etc, a simple fix

palisadoes commented:
That's the reason for this. But with this test they can see the failure and close it without one of us having to tell them to do it.

It's to relieve the volunteer workload.

It won't prevent other tests from running due to it's location in the YAML for

palisadoes commented:
Please fix the failing test

Anubhav-2003 commented:
@palisadoes Respected sir, I just wanted to ask that the first failed case was for change of files in the workflows folder. The second one was, because the present base branch does not contains the script yet. The last case was for code coverage. But most of the lines shown to be missing coverage are for other files, which I did not make a single change to. Do I need to write test cases for my script and the changes in the yaml file. Because I was unable to find any file for those tests. 

Thank you.

palisadoes commented:
Thanks. I'll merge this and if it fails on future PRs I'll ask you to fix it

Anubhav-2003 commented:
@palisadoes Thank you sir. I will be always ready for that.
